flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/StringToBinaryCastRule.java

```diff
@@ -45,6 +49,14 @@ public String generateExpression(
             String inputTerm,
             LogicalType inputLogicalType,
             LogicalType targetLogicalType) {
-        return methodCall(inputTerm, "toBytes");
+        int inputLength = LogicalTypeChecks.getLength(inputLogicalType);
+        int targetLength = LogicalTypeChecks.getLength(targetLogicalType);
+
+        if (targetLength >= inputLength) {
+            return methodCall(inputTerm, "toBytes");
+        } else {
+            return staticCall(
+                    Arrays.class, "copyOfRange", methodCall(inputTerm, "toBytes"), 0, targetLength);
+        }
```

Comment:
Same comment as above about the unnecessary copy, but because you invoke `toBytes()` which is potentially expensive, convert this rule to a `AbstractNullAwareCodeGeneratorCastRule` and save the `toBytes` result in a local variable